ignite/utils.py

```diff
@@ -90,6 +92,130 @@ def _tree_map(
+    tensors with special formatting rules. It's particularly useful for logging,
+    debugging, or any scenario where a human-readable representation of complex,
+    nested data structures is required.
+
+    The function handles the following types:
+
+    - Numbers: Formatted to four decimal places.
+    - PyTorch tensors:
+        - Scalars are formatted to four decimal places.
+        - 1D tensors with more than 10 elements show the first 10 elements
+          followed by an ellipsis.
+        - 1D tensors with 10 or fewer elements are fully listed.
+        - Multi-dimensional tensors display their shape.
+    - Dictionaries: Each key-value pair is included in the output with the key
+      as a prefix.
+    - Lists and tuples: Flattened and included in the output. Empty lists/tuples are represented
+      by an empty string.
+    - None values: Represented by an empty string.
+
+    Args:
+        data: The input data to be flattened and formatted. It can be a nested
+            combination of dictionaries, lists, tuples, numbers, and PyTorch
+            tensors.
+
+    Returns:
+        A list of formatted strings, each representing a part of the input data
+        structure.
+
+    Examples:
+        Use the function to flatten and format various data structures.
+
+        .. testsetup:: *
+
+            from ignite.utils import flatten_to_str
+            import torch
+
+        .. testcode::
+
+            flatten_to_str(42)
+
+        .. testoutput::
+
+            ['42.0000']
+
+        .. testcode::
+
+            flatten_to_str([1, 2.34567, 3])
+
+        .. testoutput::
+
+            ['1.0000, 2.3457, 3.0000']
+
+        .. testcode::
+
+            flatten_to_str({'a': 1, 'b': [2, 3], 'c': None})
+
+        .. testoutput::
+
+            ['a: 1.0000', 'b: 2.0000, 3.0000', 'c: ']
+
+        .. testcode::
+
+            flatten_to_str(torch.tensor([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]))
+
+        .. testoutput::
+
+            ['[1.0000, 2.0000, 3.0000, 4.0000, 5.0000, 6.0000, 7.0000, 8.0000, 9.0000, 10.0000, ...]']
+
+        .. testcode::
+
+            flatten_to_str({'tensor': torch.tensor([[1, 2], [3, 4]])})
+
+        .. testoutput::
+
+            ['tensor: Shape: torch.Size([2, 2])']
+    """
+    formatted_items: List[str] = []
+
+    def format_item(item: Any, prefix: str = "") -> Optional[str]:
+        if isinstance(item, numbers.Number):
+            return f"{prefix}{item:.4f}"
+        elif torch.is_tensor(item):
+            if item.dim() == 0:
+                return f"{prefix}{item.item():.4f}"  # Format scalar tensor without brackets
+            elif item.dim() == 1 and item.size(0) > 10:
+                return f"{prefix}[" + ", ".join(f"{x.item():.4f}" for x in item[:10]) + ", ...]"
+            elif item.dim() == 1:
+                return f"{prefix}[" + ", ".join(f"{x.item():.4f}" for x in item) + "]"
+            else:
+                return f"{prefix}Shape: {item.shape}"
```

Comment:
```suggestion
                return f"{prefix}Shape: {list(item.shape)}"
```
to avoid unnecessary `torch.Size(` substrings. Maybe, even "Shape:" can be omitted to save space.

Reply:
But we say "Shape" before providing the shape values, so this is clear. Actually, I wonder if this finally makes sense even to log the shape in the loop...

```
2024-04-22 12:05:47,843 trainer INFO: Train: start epoch [1/4]
... Epoch [1/4]  [20/100]:  ETA: 0:00:00  lr: 0.00100  ...  Shape: [32, 3, 256, 256]
... Epoch [1/4]  [40/100]:  ETA: 0:00:00  lr: 0.00100  ...  Shape: [32, 3, 256, 256]
... Epoch [1/4]  [60/100]:  ETA: 0:00:00  lr: 0.00100  ...  Shape: [32, 3, 256, 256]
... Epoch [1/4]  [80/100]:  ETA: 0:00:00 lr: 0.00100   ...  Shape: [32, 3, 256, 256]
```